src/ffmpeg_subscriber.cpp

```diff
@@ -51,66 +83,131 @@ void FFMPEGSubscriber::subscribeImpl(
-      RCLCPP_DEBUG_STREAM(logger_, "was previously declared: " << kv.first);
-      v = node->get_parameter(key).get_parameter_value();
+      RCLCPP_DEBUG_STREAM(logger_, "was previously declared: " << p_name);
+      decoders = node_->get_parameter_or<std::string>(key, "");
     }
-    if (v.get<std::string>() != kv.second) {
-      RCLCPP_INFO_STREAM(
-        logger_, "overriding default decoder " << kv.second << " for " << kv.first << " with "
-                                               << v.get<std::string>());
+    if (!decoders.empty()) {
+      break;
     }
   }
-  const bool mp = ffmpeg_encoder_decoder::get_safe_param<bool>(
-    node_, param_namespace_ + "measure_performance", false);
-  decoder_.setMeasurePerformance(mp);
+  return (decoders);
 }
 
-void FFMPEGSubscriber::internalCallback(const FFMPEGPacketConstPtr & msg, const Callback & user_cb)
+void FFMPEGSubscriber::handleAVOptions(const std::string & opt)
 {
-  if (!decoder_.isInitialized()) {
-    if (msg->flags == 0) {
-      return;  // wait for key frame!
-    }
-    if (msg->encoding.empty()) {
-      RCLCPP_ERROR_STREAM(logger_, "no encoding provided!");
-      return;
+  const auto split = split_by_char(opt, ',');
+  for (const auto & sl : split) {
+    const auto kv = split_by_char(sl, ':');
+    if (kv.size() != 2) {
+      RCLCPP_WARN_STREAM(logger_, "skipping bad AV option: " << sl);
+    } else {
+      decoder_.addAVOption(kv[0], kv[1]);
+      RCLCPP_INFO_STREAM(logger_, "setting AV option " << kv[0] << " to " << kv[1]);
     }
-    userCallback_ = &user_cb;
-    const std::string decoder = ffmpeg_encoder_decoder::get_safe_param<std::string>(
-      node_, param_namespace_ + msg->encoding, "");
-    if (decoder.empty()) {
-      RCLCPP_ERROR_STREAM(logger_, "no valid decoder found for encoding: " << msg->encoding);
-      return;
+  }
+}
+
+void FFMPEGSubscriber::internalCallback(const FFMPEGPacketConstPtr & msg, const Callback & user_cb)
+{
+  if (decoder_.isInitialized()) {
+    // the decoder is already initialized
+    decoder_.decodePacket(
+      msg->encoding, &msg->data[0], msg->data.size(), msg->pts, msg->header.frame_id,
+      msg->header.stamp);
+    return;
+  }
+  // need to initialize the decoder
+  if (msg->flags == 0) {
+    return;  // wait for key frame!
+  }
+  if (msg->encoding.empty()) {
+    RCLCPP_ERROR_STREAM(logger_, "no encoding provided!");
+    return;
+  }
+  userCallback_ = &user_cb;
+  const auto codec = split_by_char(msg->encoding, ';')[0];
+  std::string decoder_names = getDecodersFromMap(msg->encoding);
+  decoder_names = ffmpeg_encoder_decoder::utils::filter_decoders(codec, decoder_names);
+  if (decoder_names.empty()) {
+    decoder_names = ffmpeg_encoder_decoder::utils::find_decoders(codec);
+    RCLCPP_WARN_STREAM(
+      logger_, "no decoders configured for encoding " << msg->encoding
+                                                      << " defaulting to: " << decoder_names);
+  }
+  if (decoder_names.empty()) {
```

Comment:
@berndpfrommer should this be right after `find_decoders`, before logging the `defaulting to ...` warning?